src/ngraph/ngraph_imperative.cc

```diff
@@ -87,69 +87,128 @@ void NGImperative::parse_ngraph() {
     }
 }
 
+bool compute_forward_imperative(const nnvm::NodeAttrs &attrs,
+                                const mxnet::OpContext &ctx,
+                                const std::vector<mxnet::NDArray> &inputs,
+                                const std::vector<mxnet::OpReqType> &req,
+                                const std::vector<mxnet::NDArray> &outputs) {
+  std::shared_ptr<Graph> op_ng;
+  int mode = static_cast<int>(GraphExeMode::kInfer);
+  if (!sparse_check(inputs) && !sparse_check(outputs)) {
+    // thread local cache for ngraph op
+    static thread_local NGIOpCache ngicache;
+    auto op_key = get_ngiop_key(attrs, ctx.run_ctx.ctx, inputs);
+    op_ng = ngicache[op_key];
+    if (!op_ng) {
+      NGImperative ngi(attrs, ctx.run_ctx.ctx, inputs, &req, outputs);
+      op_ng = ngicache[op_key] = ngi.get_op_ngraph();
+      if (ngraph_log_verbose && op_ng)
```

Comment:
Two things:

(1) Style suggestion: `if` statements that are _not_ followed by braces cause billions of gruesome deaths world-wide _every single day_.

(2) Is it actually possible for `op_ng` to be null at this point?

Reply:
agree about if statements, ngraph seems to enforce this very well. op_ng can be null if we cant successfully create ngraph IR.

Reply:
> op_ng can be null if we cant successfully create ngraph IR.

Under what circumstances can that happen?  I'm asking to confirm that we're handling errors in whatever way(s) we intend.

Reply:
mainly if ngraph sg_compiler couldnt compile (eg dtype not supported); Here we are ensuring op_ng is not null before dispatching to ngraph, else we have a fallback from mxnet that we can dispatch to.